### ISSUE TITLE
Adding scale parameter to losses

### ISSUE THREAD

**Issue by ahwillia** [JuliaML/Losses.jl#27]
Suppose I'm working on a dataset with `n` observations of `p` measured features. For simplicity let's say all measurements are real-valued and I want to use `L2DistLoss`. However, each feature has a difference variance, which could happen for a variety of reasons.

As a result, I want to scale the loss of each feature so that errors in the noisier features are penalized less. For example, if `p = 2` I might calculate the full objective function like this:

```julia
for n = 1:nobs
    # penalty for first feature with scale 1.0
    objective += 1.0*value(L2DistLoss(), actual[n,1] - prediction[n,1])
    # penalty for second feature with scale 2.0
    objective += 2.0*value(L2DistLoss(), actual[n,2]  - prediction[n,2])
end
```

My question is whether it would be useful to define a new type called something like `ScaledLoss` so that the calculation becomes:

```julia
for n = 1:nobs
    # penalty for first feature with scale 1.0
    objective += value(ScaledLoss{L2DistLoss,1.0}(), actual[n,1] - prediction[n,1])
    # penalty for second feature with scale 2.0
    objective += value(ScaledLoss{L2DistLoss,2.0}(), actual[n,2]  - prediction[n,2])
end
```

The advantage being that this scaling can be passed through when calculating gradients, etc.

**Note:**  these loops are really just for demonstration. I wouldn't actually calculate the objective this way, but I think it illustrates the point.

**Comment by Evizero** [comment]
Never seen such a use-case. Usually one has some kind of feature reducing/combining function, like a linear predictor `prediction = feature1 * data1 + feature2 * data2`. So one usually does not have one prediction per feature or something of that sorts. 

That said, I don't see a reason not to offer it if it is useful to you. Do you have a practical example of something like this?

**Comment by ahwillia** [comment]
Factor analysis would be a simple example. This is basically how [LowRankModels.jl](https://github.com/madeleineudell/LowRankModels.jl) implements losses.

Basically any time you have a **multiple predicted features** this might be useful. Your example just has one so the scaling doesn't matter. But if you had `prediction1` and `prediction2` for each observation, you might want to penalize errors on one more harshly than the other. E.g. it is more critical for one prediction to be accurate than the other for your application.

**Comment by Evizero** [comment]
I see. I stand corrected then. Looks like a good idea then. Not sure that the scale factor needs to be a part of the type itself though. Does this give you any benefits ?

**Comment by Evizero** [comment]
I just looked at your branch. seems like you did not actually include it into the type after all. I retract my question :)

**Comment by tbreloff** [comment]
I've used this sort of thing before.  I think it would be an alternative to distance losses.  Shouldn't it operate on arrays? Not scalars.

**Comment by ahwillia** [comment]
I think the most flexible is for it to operate on scalars, since you might want different losses (as well as different scalings). For example, let's say I have 3 predicted values, the first is binary so I want `LogisticLoss`, the second has outliers so I want to use `L1DistLoss`, and the third is `L2DistLoss`. Then I could represent the total objective function as an array:

```julia
losses = [ ScaledLoss(LogisticLoss,1.0), ScaledLoss(L1DistLoss,2.0), ScaledLoss(L2DistLoss,1.5) ]
```

(I added some arbitrary scalings for good measure.) An alternative would be to have two different arrays:

```julia
losses = [ LogisticLoss(), L1DistLoss(), L2DistLoss() ]
scalings = [ 1.0, 2.0, 1.5 ]
```

I'd rather not have two different arrays floating around in my code that always depend on each other. So I guess in the second case I would prefer to collect them into a type.

Any preferences on this? I was thinking the first option, along the lines of:

```julia
immutable ScaledLoss{L<:Loss,T<:Number} <: Loss
    loss::L
    λ::T
end

value(l::ScaledLoss, r::Number) = l.λ*value(l.loss,r)
deriv(l::ScaledLoss, r::Number) = l.λ*deriv(l.loss,r)
```

**Comment by Evizero** [comment]
could we not overload `*` in such way that `2 * L1DistLoss()` would perform `ScaledLoss(L1DistLoss(), 2)` for us, instead of having to call the verbose constructor ourselves ?

**Comment by tbreloff** [comment]
I don't love it, mainly because it seems like you have to be incredibly verbose to do simple things.  If L1DistLoss can operate on vectors, what's the equivalent for scaled loss?

**Comment by Evizero** [comment]
@tbreloff the array logic is taken care of here https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/supervised.jl Each loss itself needs only implement the scalar logic

**Comment by tbreloff** [comment]
Doesn't that API break with this proposal?  It assumes a consistent loss for all array elements.

**Comment by Evizero** [comment]
I don't see why it would, but that may just be because I understand the use-case too little. To me it looks like the idea is that the total cost is a (weighted) sum of a number of losses, which would to me just mean that one calls `cost = value(loss1, ...) + value(loss2, ...) + value(loss3,...) + ....`. So for each call of `value` itself the loss would remain consistent. Am I wrong in that assumption?

**Comment by Evizero** [comment]
One thing I fail to see is how scaling a loss would change the end outcome. Doesn't the minimum stay at the exact same place? I guess scaling the derivative would mean faster/slower movement but in the end the minimizer should be independent of the scale factor, no?

**Comment by tbreloff** [comment]
No... it would be (I think):

```
total_loss = sum(abs2(w .* diff))
# or
total_loss = sum(w .* abs2(diff))
```

where diff is a vector of differences, and w is a vector of scales/weightings.

**Comment by Evizero** [comment]
i get that the total loss is different, but the value of the loss isn't that meaningful. the parameters producing `diff` that minimize the loss  in your example  should be the same as for `sum(abs(diff))` though.

**Comment by tbreloff** [comment]
Maybe for simple models?  That assumption is wrong for a ANN with one hidden layer.

**Comment by tbreloff** [comment]
The weights from input to the hidden layer impact ALL of the output values, and are a weighted sum of all the loss components.

**Comment by Evizero** [comment]
Oh, wait. I think I am catching on (I am furiously scribbling on my whiteboard the whole time). The point is not to create a scaled loss as an alternative the the old one, it is to weigh importance to influencing factors. So while a scaled loss by itself isn't that meaningful, it does make sense when combining it with other scaled losses.

**Comment by Evizero** [comment]
@tbreloff but the weights you are talking about are parameters of the prediction function, they are something completely different and not scaling factors to the loss itself.

The `w` in this code from above are not neural network weights, it is just some factor used to scale the loss after it was computed.

```
total_loss = sum(abs2(w .* diff))
```

because `diff = prediction - actual`. The neural network weights were used to produce `prediction`.

Am I missing the point? I feel like I am missing the point and am making a fool out of myself right now.

**Comment by tbreloff** [comment]
I think you'd agree that we have a loss so that we know how to update learnable parameters of the model which output the prediction, correct?  In that case we're choosing scaling factors on each output difference (or squared diff) in order to adjust our model parameters to "get some outputs really right at the expense of others".

**Comment by Evizero** [comment]
```julia
immutable ScaledLoss{L<:Loss,T<:Number} <: Loss
    loss::L
    λ::T
end
```

This code from above looks to me like there is one single scalar factor for the scaled loss. The argument you are making on the other hand reads to me like it builds on the premise that I weigh different observations with a different scaling factor.

Regardless, I think I now understand that scaling a loss makes sense iff I use it in combination with other scaled losses. I still believe that if used just by itself a scaled version of an L2DistLoss will have the same minimizer as the unscaled version, but that is not relevant here.

**Comment by tbreloff** [comment]
>  if used just by itself a scaled version of an L2DistLoss will have the same minimizer as the unscaled version

Correct.  Pointless.  This only matters if you have different scaling factors on outputs.

**Comment by Evizero** [comment]
Fun fact. The `ModifiedHuberLoss` (which is implemented) is actually a scaled version of a special case of the `SmoothedL1HingeLoss` (which is also implemented) and thus kinda redundant. I did implement it anyway because it is famous enough that people would want to plot it or use it.

**Comment by ahwillia** [comment]
Woah, I missed a lot already. @Evizero - I think you understand the gist of it correctly.

@tbreloff  brings up a very good point:

> Doesn't that API break with this proposal? It assumes a consistent loss for all array elements.

I think you're right since I specified `r::Number`. Good catch, but wouldn't this fix it? Or would this somehow be ambiguous with the existing API?

```julia
value(l::ScaledLoss, r::AbstractArray) = l.λ*value(l.loss,r)
deriv(l::ScaledLoss, r::AbstractArray) = l.λ*deriv(l.loss,r)
```

**Comment by Evizero** [comment]
@ahwillia if `l.λ` is a scalar and if you want to apply `value` elementwise (so no sum or reduction)  then you don't need to implement a `value` for arrays, because that is taken care of

**Comment by ahwillia** [comment]
The problem is that I want to make sure I multiply by `l.λ` **after** the reduction operation of `l.loss`.

**Comment by Evizero** [comment]
the output of `value(l.loss,r)` has the same dimension as `r`. no reduction here

**Comment by Evizero** [comment]
This should be the callstack basically

1. https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/supervised.jl#L39-L42
2. https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/supervised.jl#L56-L63
3. https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/supervised.jl#L204

**Comment by Evizero** [comment]
oh I now see we don't have a distributor function for when `r` is an array. My bad then. Could be implemented similar to 1. though

**Comment by ahwillia** [comment]
Oh, I misunderstood... Why don't we have something like `value_reduce(loss,actual,predicted)` that returns a single number when actual and predicted are two arrays of matching size?

> we don't have a distributor function for when r is an array.

+1 for adding this.

**Comment by Evizero** [comment]
> Why don't we have something like `value_reduce(loss,actual,predicted)` that returns a single number when actual and predicted are two arrays of matching size?

we have (two actually)

- https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/supervised.jl#L128
- https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/supervised.jl#L150

**Comment by Evizero** [comment]
> +1 for adding this.

I agree, I actually was confident we actually have this already. Can't think of a good reason why we shouldn't. I think I just overlooked it at some point

**Comment by Evizero** [comment]
I just looked at `sumvalue` and `meanvalue` a little closer. They seriously need some revision concerning type stability *note_to_self*.

**Comment by ahwillia** [comment]
> we have (two actually)

😅 now I'm embarrassing myself. So I guess what I want is:

```julia
sumvalue(l::ScaledLoss, target::AbstractArray, output::AbstractArray) = l.λ*sumvalue(l.loss,target,output)
grad(l::ScaledLoss, target::AbstractArray, output::AbstractArray) = l.λ*grad(l.loss,target,output)
```

Implementing `value` and `deriv` directly makes little sense for a scaled loss. I think this may have been the source of your confusion @Evizero. Could we just leave `value`, `deriv`, etc. undefined?

**Side note:** why is `target` constrained to be `AbstractVector` and not `AbstractArray` in sumvalue and meanvalue?

**Comment by Evizero** [comment]
> Could we just leave value, deriv, etc. undefined?

We should still define them for completeness' sake. Scaled losses might be redundant, but the are not equivalent to their unscaled counter parts. The learning behaviour would differ for example (if one doesn't normalize the gradient).

> Side note: why is target constrained to be AbstractVector and not AbstractArray in sumvalue and meanvalue?

I remember having some reason for that, but I do not remember the reason. I will revisit those functions.

**Comment by Evizero** [comment]
I also would like there to be (untested)

```
Base.*(factor::Number, loss::Loss) = ScaledLoss(loss, factor)
```

**Comment by Evizero** [comment]
> Side note: why is target constrained to be AbstractVector and not AbstractArray in sumvalue and meanvalue?

Ah, I think I remember. I wanted to implement those function for vectors only, because I was yet unsure who to treat multivariate losses. So both parameter where thought of as losses, but I wanted to allow the output to be either row or column vectors for some usecase of mine. Anyway, I will revisit them

**Comment by ahwillia** [comment]
+1 for overloading `Base.*` - do you want me to open a PR on all this, or do you want to take the lead since you understand the codebase better?

**Comment by Evizero** [comment]
depends on your timeline. I think I can get it done before monday (maybe sooner). if that is early enough for you than I will gladly take care of this

**Comment by ahwillia** [comment]
Works for me. I have a hacky workaround in my code which will hold me over until next Monday. Thanks for this!

**Comment by Evizero** [comment]
Almost done.

```
Test Summary: |    Pass    Total
tst_loss.jl   | 2637504  2637504
```

Maybe we are overdoing the testing? :)

**Comment by ahwillia** [comment]
Closed by #29
